File: src/api/teacher.ts
import { api } from './client'
import type { Lesson, LessonStudent, SubmitGradesPayload } from '../types/lesson'

export async function getTeacherLessons(params: {
  weekOffset: number
  dayIndex: number
}): Promise<Lesson[]> {
  const { data } = await api.get<Array<{
    id: string
    subject: string
    class_id: string
    class_name: string
    time: string
    room?: string
    topic?: string | null
    homework_text?: string | null
  }>>('/teacher/lessons', { params: { week_offset: params.weekOffset, day_index: params.dayIndex } })
  return data.map((l) => ({
    id: l.id,
    subject: l.subject,
    classId: l.class_id,
    className: l.class_name,
    time: l.time,
    room: l.room,
    topic: l.topic ?? null,
    homeworkText: l.homework_text ?? null,
  }))
}

export async function getLessonStudents(lessonId: string): Promise<LessonStudent[]> {
  if (!lessonId) return []
  const { data } = await api.get<Array<{
    student_id: string
    name: string
    attendance: string
    grade: number | null
  }>>(`/teacher/lessons/${lessonId}/students`)
  return data.map((s) => ({
    studentId: s.student_id,
    name: s.name,
    attendance: s.attendance as 'present' | 'absent',
    grade: s.grade as LessonStudent['grade'],
  }))
}

export async function submitGrades(payload: SubmitGradesPayload): Promise<{ success: true }> {
  await api.post('/teacher/lessons/grades', {
    lesson_id: payload.lessonId,
    entries: payload.entries.map((e) => ({
      student_id: e.studentId,
      attendance: e.attendance,
      grade: e.grade,
    })),
    topic: payload.topic ?? null,
    homework_text: payload.homeworkText ?? null,
  })
  return { success: true }
}

File: backend/app/schemas/teacher.py
from pydantic import BaseModel
from typing import Literal


class LessonResponse(BaseModel):
    id: str
    subject: str
    class_id: str
    class_name: str
    time: str
    room: str | None = None
    topic: str | None = None
    homework_text: str | None = None


class LessonStudentResponse(BaseModel):
    student_id: str
    name: str
    attendance: Literal["present", "absent"]
    grade: int | None = None  # 2,3,4,5


class SubmitGradesEntry(BaseModel):
    student_id: str
    attendance: Literal["present", "absent"]
    grade: int | None = None


class SubmitGradesRequest(BaseModel):
    lesson_id: str
    entries: list[SubmitGradesEntry]
    topic: str | None = None
    homework_text: str | None = None


class JournalSubjectOption(BaseModel):
    id: str
    name: str


class JournalDataResponse(BaseModel):
    class_id: str
    class_name: str
    subject: str
    subject_id: str = ""  # for filtering and saving grades
    # list of subjects for class (for dropdown)
    subjects: list[JournalSubjectOption] = []
    dates: list[str]
    students: list[dict]  # id, name
    # student_id -> date -> grade
    grades: dict[str, dict[str, int | str | None]]


class SaveGradeRequest(BaseModel):
    class_id: str
    student_id: str
    date_iso: str
    value: int | str | None  # 2,3,4,5,'Н',null
    subject_id: str | None = None  # required for progress to show by subject
